core/include/detray/tools/surface_factory.hpp

```diff
@@ -0,0 +1,339 @@
+                    typename detector_t::surface_container &surfaces,
+                    typename detector_t::transform_container &transforms,
+                    typename detector_t::mask_container &masks,
+                    typename detector_t::geometry_context ctx = {}) const
+        -> dindex_range override {
+        using surface_t = typename detector_t::surface_type;
+        using mask_link_t = typename surface_t::mask_link;
+        using material_link_t = typename surface_t::material_link;
+
+        // The material will be added in a later step
+        constexpr auto no_material = surface_t::material_id::e_none;
+        // In case the surfaces container is prefilled with other surfaces
+        std::size_t surfaces_offset = surfaces.size();
+
+        // Nothing to construct
+        if (size() == 0UL) {
+            return {surfaces_offset, surfaces_offset};
+        }
+
+        for (const auto [idx, comp] : detray::views::enumerate(m_components)) {
+
+            // Add transform
+            transforms.push_back(m_transforms[idx], ctx);
+
+            // Build the given mask type and add it to the correct mask coll.
+            if constexpr (std::is_same_v<mask_shape_t, annulus2D<>>) {
+                add_annulus_mask(m_volume_link[idx], comp, masks);
+            } else if constexpr (std::is_same_v<mask_shape_t, cylinder2D<>>) {
+                add_cylinder_mask(m_volume_link[idx], comp, masks);
+            } else if constexpr (std::is_same_v<mask_shape_t, line<>>) {
+                add_line_mask(m_volume_link[idx], comp, masks);
+            } else if constexpr (std::is_same_v<mask_shape_t, rectangle2D<>>) {
+                add_rectangle_mask(m_volume_link[idx], comp, masks);
+            } else if constexpr (std::is_same_v<mask_shape_t, ring2D<>>) {
+                add_disc_mask(m_volume_link[idx], comp, masks);
+            } else if constexpr (std::is_same_v<mask_shape_t, single3D<>>) {
+                add_single_mask(m_volume_link[idx], comp, masks);
+            } else if constexpr (std::is_same_v<mask_shape_t, trapezoid2D<>>) {
+                add_trapezoid_mask(m_volume_link[idx], comp, masks);
+            } else if constexpr (std::is_same_v<mask_shape_t, unmasked>) {
+                add_unmasked(m_volume_link[idx], comp, masks);
+            }
+
+            // Add surface with all links set (relative to the given containers)
+            mask_link_t mask_link{mask_id, masks.template size<mask_id>() - 1};
+            material_link_t material_link{no_material, dindex_invalid};
+            surfaces.emplace_back(transforms.size(ctx) - 1, mask_link,
+                                  material_link, volume.index(), dindex_invalid,
+                                  sf_id);
+        }
+
+        return {surfaces_offset, surfaces.size()};
+    }
+
+    private:
+    /// @brief Build and add a stereo annulus mask to the given mask collection.
+    void add_annulus_mask(const dindex &vol_link,
```

Comment:
Is it totally impossible to have one `add_mask()` function?

``` c++
void add_mask(const dindex &vol_link, const std::vector<scalar_t>& components,  
                         typename detector_t::mask_container &masks) const {
  masks.template emplace_back<mask_id>(empty_context{}, vol_link, components);
}
```

Reply:
I added the corresponding constructor. The code is much shorter now :)